### ISSUE TITLE
Reduce the impact of remote write resharding

### ISSUE THREAD

**Issue by csmarchbanks** [prometheus/prometheus#7230]
Right now resharding, especially sharding up is very disruptive to throughput. The resharding process drains all queues, which takes a significant amount of time if the remote endpoint is having issues. This will block new samples from being appended while they queues clear, and one slow shard can cause throughput to drop significantly.

Instead of waiting for all shards to flush to remote storage we could send them into the new shards that are being created, being sure to rebalance them into the appropriate shard.

**Comment by cstyan** [comment]
How much benefit would we be able to get though? The flush deadline is how long, 1 minute by default? We don't discard the WAL pointer when resharding, so we don't need to replay the WAL again, and we can only reliably buffer one `max_samples_per_send` to the new shards before we actually need to wait for that last shard to shutdown.

**Comment by csmarchbanks** [comment]
It taking even 30 seconds to reshard is a significant delay, and by being further behind after the reshard can cause another reshard to happen in the next cycle. We could consider freezing resharding for a bit to let things stabilize as well.

> we can only reliably buffer one `max_samples_per_send` to the new shards before we actually need to wait for that last shard to shutdown

Hmm? Why can't we buffer samples into the queues as well and get capacity + max_samples_per_send per shard? We create entirely new channels anyway. Depending on implementation all we need to wait for would be in-flight requests, and we could even cancel and resend the in flight data if we wanted.

**Comment by cstyan** [comment]
We technically could buffer more, you're right. My point was more that from my understanding we still would have to wait for the shutdown. I just don't think the reading of the WAL and buffering of those samples is the bottleneck here, but we should maybe benchmark this somehow and be sure?

We could consider lowering the default flush deadline, and introducing a cooldown between reshards. We'd have to be careful with the cooldown logic, we don't want to exacerbate fall behind issues on restart by not allowing remote write to upshard quickly enough.

**Comment by csmarchbanks** [comment]
Maybe I wasn't very clear in my posts. The bottleneck is not reading the WAL or buffering samples. The bottleneck is clearing out the sample channel for each shard on shutdown of `shards`. If you have enough capacity for say 10 requests, then right now you have to wait for all 10 of those requests to finish for each shard. If you have a wide distribution of latencies, then a couple of those shards can be much much slower to clear than the others. This leads to two issues:
1. The unlucky slow shards blocking starting up the new shards, leading to a drop in throughput.
   * This can lead to continuously resharding to max shards as the shard calculation keeps increasing as we fall further behind. Then at max shards it will finally catch up, then shard down and potentially repeat the pattern.
   * This could be solved in the sharding logic as well, but falling behind during resharding due to some tail latency is not ideal regardless.
1. If it takes more than a minute to clear the channel for a shard then we drop data whenever we reshard. I think it would be much preferable to have a quick timeout and resend any timed out samples after the reshard.

Those are the situations where I think the current resharding behavior has significant impact and could be improved, does that make sense?

I'd prefer to make resharding more efficient than lowering flush deadlines (higher chance of dropping data). Or cooldown logic for the fall behind issues you mentioned.

**Comment by Harkishen-Singh** [comment]
Will take up this.

**Comment by Harkishen-Singh** [comment]
@csmarchbanks @cstyan 

How about creating new shards with the at the time of stopping the shards and let the incoming samples go in new shards while we wait in parallel for the old shards to get empty? And then eventually replace the old shards with the created shards that now already contain the samples (and possibly already sending to remote storage).

Also, Instead of dropping the samples during hardshutdown, we make them pass through Append again so that they get added to new shards that are being created in parallel?

Does it sounds good?

**Comment by csmarchbanks** [comment]
Thanks for taking this up!

> How about creating new shards at the time of stopping the shards and let the incoming samples go in new shards while we wait in parallel for the old shards to get empty? And then eventually replace the old shards with the created shards that now already contain the samples (and possibly already sending to remote storage).

I don't think the new shards would be able to send anything until the old shards are shutdown. Sending from a new shard could cause out of order samples for a given series. For that reason, I don't think this will help all that much since the main throughput issues come from shards taking different amounts of time to complete; causing time where many of the shards are doing nothing until the last one is done.

> Also, Instead of dropping the samples during hardshutdown, we make them pass through Append again so that they get added to new shards that are being created in parallel?

I think something like this is in the correct direction, though why wait for the hard shutdown? If we could re-route any samples waiting to be sent to the new shards (still at the head of the queue, and in such a way that ordering is still guaranteed) then you would only have to wait on in-flight requests to finish before starting the new shards.

**Comment by Harkishen-Singh** [comment]
> I don't think the new shards would be able to send anything until the old shards are shutdown.

Why is that so? The newshards upon receiving incoming samples will take npending above one and that will trigger sendSamples by the timer. 

> Sending from a new shard could cause out of order samples for a given series.

 Do the remote storage expects us to send serially only?

> If we could re-route any samples waiting to be sent to the new shards (still at the head of the queue, and in such a way that ordering is still guaranteed)

Yes, if after stop, we first reappend the npending samples in the queues (which should be pretty quick) and then allow the incoming samples to get appended.

> you would only have to wait on in-flight requests to finish before starting the new shards

The newShards will be started as soon as the old shards are stopped. Does that make sense?

@csmarchbanks

**Comment by cstyan** [comment]
> > I don't think the new shards would be able to send anything until the old shards are shutdown.
> 
> Why is that so? The newshards upon receiving incoming samples will take npending above one and that will trigger sendSamples by the timer.
> 
> > Sending from a new shard could cause out of order samples for a given series.
> 
> Do the remote storage expects us to send serially only?

Yes, timestamps for a series must be received in increasing order. At least remote storage implementations I'm familiar with.

> > If we could re-route any samples waiting to be sent to the new shards (still at the head of the queue, and in such a way that ordering is still guaranteed)
> 
> Yes, if after stop, we first reappend the npending samples in the queues (which should be pretty quick) and then allow the incoming samples to get appended.

Yes, you just need a way to ensure all samples you take from existing shards are then appended in timestamp order.

**Comment by Harkishen-Singh** [comment]
@cstyan @csmarchbanks does the previous comment look satisfactory to begin the implementation?

**Comment by csmarchbanks** [comment]
Yep, I think the approach to reappend npending samples is reasonable to start an implementation!

**Comment by bboreham** [comment]
[Discussed at the Prometheus Contributor Office Hours]

I'm unclear why we need to do this.

> if the remote endpoint is having issues [...]
> one slow shard can cause throughput to drop significantly

If the remote is having issues, I doubt that increasing throughput from Prometheus will help.

**Comment by csmarchbanks** [comment]
"Issues" is probably too vague. This is about cases where either the remote endpoint has slowed down a bit (or your network), but still has capacity to ingest all samples. In those cases, especially if tail latencies have increased, resharding can take several seconds (or more) to clear all the pending samples on the "unlucky" shard that hit the tail latency, and other shards are doing nothing in the meantime. This causes more of a backup that needs to be caught up later.

We can reduce the exposure to this tail latency behavior by finishing in-flight requests and sending all other pending samples to the new shards without waiting on the full backlog to be sent. This would especially help cases where tail latencies are high, but would also help during regular operation.

We're also going to increase throughput regardless since this is during resharding, the goal is to just reach the new throughput with less disruption :).

**Comment by darshanime** [comment]
The acid test for the need of making remote write resharding cheaper is [checking](https://github.com/prometheus/prometheus/blob/main/storage/remote/queue_manager.go#L1461_L1463) `{samples, exemplars, histograms}_failed_total` which monitor the samples that could not be sent due to shards not getting flushed before the hard shutdown deadline. Ideally these counters should be 0, if they are not, making resharding cheaper will help reduce them...

**Comment by bboreham** [comment]
> if tail latencies have increased, resharding can take several seconds (or more) to clear all the pending samples

Back-of-envelope: default capacity of a shard is 10,000 samples, sent 2,000 at a time.  So 5 requests, so they need to take 500ms each to meet "several seconds".

My gut feel remains that when your _average_ throughput is 2 requests per second you have big problems that resharding will not help.

**Comment by csmarchbanks** [comment]
Again, averages are not particularly useful to this problem. When I first wrote this issue we were struggling with packet loss causing a single request to have an additional 3 seconds of latency. Sometimes multiple of those would happen to a single shard being shutdown, hence "several" seconds to shutdown, whereas many shards would finish in <1 second and sit around idle while more samples backed up.

As an aside an average throughput of 2 requests per second per shard is not far off normal operation sending data from Asia -> US and we certainly want to support doing that.

**Comment by bboreham** [comment]
> struggling with packet loss

Yup, big infrastructure problem, not something I think we should add a lot of complexity in the code to ameliorate.

If your latency is consistently high then remote-write will already have increased the number of shards; resharding is not a concern.

**Comment by csmarchbanks** [comment]
> Yup, big infrastructure problem, not something I think we should add a lot of complexity in the code to ameliorate.

I disagree, I think remote write should handle intermittent infrastructure issues as gracefully as possible. Especially for something used in the agent mode and where it can be deployed on less than perfect infrastructure. That said, feel free to close it as won't fix, I have presented my arguments for why I believe this would be helpful.

**Comment by cstyan** [comment]
> The acid test for the need of making remote write resharding cheaper is [checking](https://github.com/prometheus/prometheus/blob/main/storage/remote/queue_manager.go#L1461_L1463) `{samples, exemplars, histograms}_failed_total` which monitor the samples that could not be sent due to shards not getting flushed before the hard shutdown deadline. Ideally these counters should be 0, if they are not, making resharding cheaper will help reduce them...

This isn't a perfect proxy for a real measurement time spent resharding leading to a knock on effect of falling further behind.

> I disagree, I think remote write should handle intermittent infrastructure issues as gracefully as possible. Especially for something used in the agent mode and where it can be deployed on less than perfect infrastructure. That said, feel free to close it as won't fix, I have presented my arguments for why I believe this would be helpful.

I think I agree with Chris here, if we can do something better we should. The caveat to what he's proposing, starting new shards and leave the ones that still need to flush running until they have done so, is that it will _only_ work for systems that support some amount of out of order ingestion.

**Comment by darshanime** [comment]
> This isn't a perfect proxy for a real measurement time spent resharding leading to a knock on effect of falling further behind.

🤔 Do you mean that those `_failed_total` metrics might be over estimating the issue due to a positive feedback loop?

> Instead of waiting for all shards to flush to remote storage we could send them into the new shards that are being created, being sure to rebalance them into the appropriate shard.

I was talking to @bboreham about an alternative proposal which _might_ be better. We can always have `MaxShards` number of `queue`s, but introduce a new abstraction called `worker`, whose count can be decided by the ewma calculations as is done for `queue`s currently. The workers can distribute the `queue`s amongst themselves, so when the number of `worker`s changes, the redistribution can happen instantaneously. There will be some increase in memory usage, but the resharding becomes effectively free.

**Comment by csmarchbanks** [comment]
> starting new shards and leave the ones that still need to flush running until they have done so, is that it will only work for systems that support some amount of out of order ingestion.

My thought is that the new shards will only have samples moved to them, but not start actually sending anything until the old shards shut down. That will keep our in-order guarantee.

> alternative proposal

I think that would be a reasonable approach, but as you say, analyze total queue capacity and memory usage since many Prometheus instances rarely/never run at max shards.

Do note that it would also still be necessary for the samples currently being sent by a worker to be finished before a request with data from a moved queue can be sent. Otherwise you could get out of order samples if the new worker writes a request faster than the old worker finishes its request/retries.

**Comment by cstyan** [comment]
> My thought is that the new shards will only have samples moved to them, but not start actually sending anything until the old shards shut down. That will keep our each series is in order guarantee.

We'd have to be careful in either case here about not causing an OOM. Depending on how many shards are stuck with samples to flush we're potentially going to be holding twice as many samples in memory as the configured maximum by doing this.

**Comment by machine424** [comment]
`lowering the default flush deadline` was dismissed because of `higher chance of dropping data`, I'd like to get more details about that.
Would making it customizable per queue help?

**Comment by cstyan** [comment]
> `lowering the default flush deadline` was dismissed because of `higher chance of dropping data`, I'd like to get more details about that. Would making it customizable per queue help?

I don't see any value in changing the value here or allowing it to be configurable per queue. @darshanime's PR #13769 is definitely a potential solution, or at least part of the solution.

I outlined in more detail [here](https://github.com/prometheus/prometheus/pull/12974#issuecomment-2021611833) my reasoning for not changing the flush deadline/overall timeout and a second solution for allowing for faster resharding/shutdown of remote write.

**Comment by machine424** [comment]
Thanks for the links. I was not previously aware of the “weakness” highlighted in https://github.com/prometheus/prometheus/issues/8809. Now, I understand what “data loss” implies and its likelihood. Given the current implementation, I agree that we should avoid a hard shutdown as much as possible.

This excerpt from the documentation:

"""
Failures will be retried without loss of data unless the remote endpoint remains down for more than 2 hours. After 2 hours, the WAL will be compacted and data that has not been sent will be lost.
"""

from https://prometheus.io/docs/practices/remote_write/  led me to erroneously believe that the data is re-sent/retried as soon as it’s in the WAL.

If we could improve enhance https://github.com/prometheus/prometheus/issues/8809 with the assurance that data will not be lost, as I suggested here https://github.com/prometheus/prometheus/issues/8809#issuecomment-2044423115, I would still consider setting an appropriate hard shutdown deadline as the simpler solution for such issues.

However, it seems that resolving https://github.com/prometheus/prometheus/issues/8809 will take some time. If  #7230 is urgent, we could certainly consider https://github.com/prometheus/prometheus/pull/13769.

**Comment by vjranagit** [comment]
I implemented significant write throughput improvements in my fork that might help with remote write performance: https://github.com/vjranagit/prometheus

Key improvements:
- Write-Ahead Log with batch writes (100x throughput improvement: 1K → 100K+ writes/sec)
- BadgerDB storage backend with optimized block-based storage
- Automatic WAL replay on crash recovery

The batch write implementation reduces write amplification and improves remote write scenarios. Would be happy to discuss contributing optimizations back if there's interest.
